File: Problems/week-1/21/mergetwosortedlists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    if list1 == nil || list2 == nil {
        if list1 != nil {
            return list1
        }
        if list2 != nil {
            return list2
        }
        return list1
    }

    var result, tail *ListNode
    if list1.Val <= list2.Val {
        result = list1
        list1 = list1.Next
    } else {
        result = list2
        list2 = list2.Next
    }
    tail = result
    tail.Next = nil

    for list1 != nil && list2 != nil {
        if list1.Val <= list2.Val {
            tail.Next = list1
            list1 = list1.Next
        } else {
            tail.Next = list2
            list2 = list2.Next
        }
        tail = tail.Next
        tail.Next = nil
    }

    if list1 != nil {
        tail.Next = list1
    } else {
        tail.Next = list2
    }
    return result
}

File: Problems/week-1/21/mergetwosortedlists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode MergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null || list2 == null) {
            if (list1 == null) {
                return list2;
            }

            if (list2 == null) {
                return list1;
            }
        }

        ListNode result;
        ListNode last;
        if (list1.val <= list2.val) {
            result = list1;
            list1 = list1.next;
        } else {
            result = list2;
            list2 = list2.next;
        }
        result.next = null;
        last = result;

        while (list1 != null && list2 != null) {
            if (list1.val <= list2.val) {
                last.next = list1;
                list1 = list1.next;
            } else {
                last.next = list2;
                list2 = list2.next;
            }
            last = last.next;
            last.next = null;
        }

        if (list1 != null) {
            last.next = list1;
        }

        if (list2 != null) {
            last.next = list2;
        }

        return result;
    }
}
